File: Python/80.py
H, W, K, V = map(int,input().split())
A = []

for i in range(H):
    a = list(map(int,input().split()))
    A.append(a)
    

box = [[0]*(W+1) for _ in range(H+1)]
for i in range(H):
    for j in range(W):
        box[i+1][j+1] = A[i][j]
for i in range(H):
    for j in range(1,1+W):
        box[i+1][j] += box[i][j]

for i in range(1,H+1):
    for j in range(W):
        box[i][j+1] += box[i][j]
ans = 0
for i in range(1,1+H):
    for j in range(1,1+W):
        con = box[i][j]
        if con + (K * ((i) * (j))) <= V:
            ans = max(ans, (i) * (j))
        for k in range(i):
            for l in range(j):
                con = box[i][j]
                con -= box[i][l]
                con -= box[k][j]
                con += box[k][l]
                if con + (K * ((i - k) * (j - l))) <= V:
                    ans = max(ans, (i - k) * (j - l))
print(ans)

File: Redcoder/C++/80.cpp
#include <bits/stdc++.h>
#define rep(i, nn, n) for (int i = nn; i < (n); ++i)
#define rrep(i, nn, n) for (int i = nn; i > (n); --i)
using namespace std;
using ll = long long;
using P = pair<int, int>;
#define chmin(x, y) x = min(x, y)
#define chmax(x, y) x = max(x, y)

int main()
{
    int H, W, K;
    ll V;
    cin >> H >> W >> K >> V;

    vector<vector<ll>> box(H + 1, vector<ll>(W + 1, 0));
    rep(i, 0, H) rep(j, 0, W) cin >> box[i + 1][j + 1];
    rep(i, 1, H + 1) rep(j, 0, W)
    {
        box[i][j + 1] += box[i][j];
    }
    rep(i, 0, H) rep(j, 1, W + 1)
    {
        box[i + 1][j] += box[i][j];
    }
    int ans = 0;
    rep(i, 0, H) rep(j, 0, W) rep(gi, i + 1, H + 1) rep(gj, j + 1, W + 1)
    {
        int m = (gi - i) * (gj - j);
        ll cost = box[gi][gj] - box[i][gj] - box[gi][j] + box[i][j] + 1LL * K * m;
        if (cost <= V)
            chmax(ans, m);
    }

    cout << ans << endl;
    return 0;
}